File: question/leetcode/leetcode0538/leetcode0538.go
package leetcode

import "clarencedu/algorithm-go/structure"

type TreeNode = structure.TreeNode

// 根据定义，dfs按照 右-根-左的顺序遍历二叉树，累加节点值，更新节点值
func convertBST(root *TreeNode) *TreeNode {
	sum := 0
	var dfs func(*TreeNode)
	dfs = func(node *TreeNode) {
		if node == nil {
			return
		}
		dfs(node.Right)
		sum += node.Val
		node.Val = sum
		dfs(node.Left)
	}

	dfs(root)
	return root
}

File: question/leetcode/leetcode0538.Convert-BST-to-Greater-Tree/leetcode0538.ts
class TreeNode {
	val: number
	left: TreeNode | null
	right: TreeNode | null
	constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
		this.val = (val === undefined ? 0 : val)
		this.left = (left === undefined ? null : left)
		this.right = (right === undefined ? null : right)
	}
}

function convertBST(root: TreeNode | null): TreeNode | null {
	let sum = 0;

	const dfs = (root: TreeNode | null) => {
		if (root === null) {
			return;
		}

		dfs(root.right);
		sum += root.val;
		root.val = sum;
		dfs(root.left);
	}

	dfs(root);
	return root;
};